src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+/// where each update is of the form `(Row, time, diff)` and each `(Row, time)` pair occurs
+/// exactly once and all diffs are nonzero.
+///
+/// Note that all `Batch`s keep their data on persistent storage. No data resides on disk
+/// (except currently we load all the data from batches into memory for compaction and on
+///  restart but this will get fixed!).
+///
+/// The `Coordinator` thread tells the `Compacter` when it needs to
+///  * start keeping track of a new relation
+///  * stop keeping track of a relation
+///  * resume keeping track of a relation with an already initialized Trace (on restart)
+///  * advance a relation's compaction frontier. Note that this doesn't automatically trigger
+///    any actual compaction. That happens later (keep reading).
+///
+/// The `Compacter` task periodically checks each relation's WAL directory to look
+/// for finished log segments, converts them to `Batch`s (basically consolidates the
+/// updates for a range of times) and adds them to the relation's `Trace`.
+///
+/// When a `Trace` contains too many `Batches`, the Trace physically
+/// combines all of them into a single large batch with updates
+/// forwarded to the compaction frontier.
+
+// TODO: Lets add some jitter to compaction so we aren't compacting every single
+// relation at the same time maybe?
+static COMPACTER_INTERVAL: Duration = Duration::from_secs(300);
+
+/// Instructions that the Coordinator sends the Compacter.
+#[derive(Debug)]
+pub enum CompacterMessage {
+    Add(GlobalId),
+    Drop(GlobalId),
+    Resume(GlobalId, Trace),
+    AllowCompaction(GlobalId, Timestamp),
+}
+
+/// A Batch contains all of the updates that originated within some time range [lower, upper)
+/// but the data live the file at `path`.
+///
+/// The data stored in each batch are triples of (Row, Timestamp, Diff) such that there is
+/// exactly one copy of each (Row, Timestamp) in each batch. Batches also have a header
+/// and a footer indicating the upper and lower bound timestamps.
+/// TODO: Batches are meant to mimic differential / cdcv2 batches but do not do so
+/// currently. Let's fix that. Specifically, introduce a `since` field, and counts
+/// for the number of updates at each time.
+#[derive(Debug)]
+struct Batch {
+    upper: Timestamp,
+    lower: Timestamp,
+    path: PathBuf,
+}
+
+impl Batch {
+    /// Create a batch from a finished log segment file.
+    fn create(log_segment_path: &Path, trace_path: &Path) -> Result<Self, anyhow::Error> {
+        let messages = read_segment(log_segment_path)?;
+        Batch::create_from_messages(messages, trace_path, None)
+    }
+
+    /// Read in and consolidate a list of messages, and write them to a new batch file
+    /// in `trace_path`.
+    ///
+    /// Will also compact updates up to `compaction_frontier` if provided..
+    fn create_from_messages(
+        messages: Vec<Message>,
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
```

Comment:
Idk if this helps at all, but you can always use `Timestamp::minimum()` as a compaction frontier to leave things as they are. It might simplify the cases in the subsequent code, or it might fail to communicate an important distinction.